### PR TITLE
Add more details to ModelAttributes

### PR DESCRIPTION
**Pull request by rynowak**
This change adds more information to ModelAttributes, so that metadata
providers can look at the attributes on the property and type separately
if so desired

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DataMemberRequiredBindingMetadataProvider.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ public void GetBindingMetadata([NotNull] BindingMetadataProviderContext context)
             }
 
             var dataMemberAttribute = context
-                .Attributes
+                .PropertyAttributes
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DataMemberRequiredBindingMetadataProvider.cs]: This is slightly more correct than what we were doing here before (which relies on the fact that `[DataMember]` can't be put on a type).

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelAttributesTest.cs [hunk 2]
```diff
@@ -112,10 +140,45 @@ public void GetAttributesForType_IncludesMetadataAttributes()
...
+            Assert.IsType<ClassValidator>(attributes.Attributes[2]);
+            Assert.IsType<BindAttribute>(attributes.Attributes[3]);
+
+            Assert.Equal(2, attributes.PropertyAttributes.Count);
+            Assert.IsType<RequiredAttribute>(attributes.PropertyAttributes[0]);
+            Assert.IsType<RangeAttribute>(attributes.PropertyAttributes[1]);
+
+            Assert.Equal(2, attributes.TypeAttributes.Count);
+            Assert.IsType<ClassValidator>(attributes.TypeAttributes[0]);
+            Assert.IsType<BindAttribute>(attributes.TypeAttributes[1]);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelAttributesTest.cs]: This (critical) test was missing. The wasn't anything verifying ordering.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs [hunk 3]
```diff
@@ -19,7 +19,7 @@ public class DefaultMetadataDetails
         /// </summary>
         /// <param name="key">The <see cref="ModelMetadataIdentity"/>.</param>
         /// <param name="attributes">The set of model attributes.</param>
-        public DefaultMetadataDetails(ModelMetadataIdentity key, IReadOnlyList<object> attributes)
+        public DefaultMetadataDetails([NotNull] ModelMetadataIdentity key, [NotNull] ModelAttributes attributes)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs]: so a `struct` can be `null` now?  :smiley_cat:

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 4]
```diff
@@ -48,7 +48,7 @@ public IReadOnlyList<object> Attributes
         {
             get
             {
-                return _details.Attributes;
+                return _details.Attributes.Attributes;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs]: `ModelAttributes` now available throughout the metadata details processing but not after.  should the whole thing also be available in `ModelMetadata`?

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs [hunk 5]
```diff
@@ -28,7 +28,7 @@ public DefaultMetadataDetails(ModelMetadataIdentity key, IReadOnlyList<object> a
         /// <summary>
         /// Gets or sets the set of model attributes.
         /// </summary>
-        public IReadOnlyList<object> Attributes { get; }
+        public ModelAttributes Attributes { get; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs]: should this be renamed `ModelAttributes`?

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelAttributes.cs [hunk 6]
```diff
@@ -10,60 +10,80 @@
...
         /// </param>
-        /// <returns>An <see cref="IEnumerable{object}"/> containing the attributes on the
-        /// <paramref name="property"/> before the attributes on the <paramref name="property"/> type.</returns>
-        public static IEnumerable<object> GetAttributesForProperty([NotNull] Type type, [NotNull] PropertyInfo property)
+        /// <returns>A <see cref="ModelAttributes"/> instance with the attributes of the property.</returns>
+        public static ModelAttributes GetAttributesForProperty([NotNull] Type type, [NotNull] PropertyInfo property)
         {
-            // Return the property attributes first.
             var propertyAttributes = property.GetCustomAttributes();
             var typeAttributes = property.PropertyType.GetTypeInfo().GetCustomAttributes();
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelAttributes.cs]: (might have come up when @kirthik originally added `ModelMetadataTypeAttribute` support but) should this handle `ModelMetadataTypeAttribute` on `property.PropertyType`?  e.g. could call `GetAttributesForType(property.PropertyType)`

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs]: heh.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs]: Moot point - `Attributes` is only defined on `DefaultModelMetadata`. So really this isn't exposed unless you downcast.

I can change this to expose `ModelAttributes` instead

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs]: sure

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelAttributes.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelAttributes.cs]: Will open a bug. This is not related to this PR.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelAttributes.cs]: Also, if you look at the test I added, it includes:
1). Attributes on the property
2). Attributes on the property of the metadata class
3). Attributes on the property's type
4). Attributes on the type of the property of the metadata class

**Comment by rynowak** [comment]
@dougbu updated

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/BindingMetadataProviderContext.cs [hunk 7]
```diff
@@ -36,6 +39,16 @@ public class BindingMetadataProviderContext
         public ModelMetadataIdentity Key { get; }
 
         /// <summary>
+        /// Gets the property attributes.
+        /// </summary>
+        public IReadOnlyList<object> PropertyAttributes { get; }
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/BindingMetadataProviderContext.cs]: In theory could avoid "extra storage" and stop copying the ref in the ctor and implement this as something like:

```c#
public IReadOnlyList<object> PropertyAttributes { get; } => attributes.PropertyAttributes;
```

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/BindingMetadataProviderContext.cs]: That would award a lot of #PranavPoints! I can't believe I didn't think of this.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DataAnnotationsMetadataProviderTest.cs [hunk 8]
```diff
@@ -551,7 +551,7 @@ public void GetValidationMetadata_RequiredAttribute_SetsIsRequiredToTrue()
 
             var attributes = new Attribute[] { required };
             var key = ModelMetadataIdentity.ForProperty(typeof(int), "Length", typeof(string));
-            var context = new ValidationMetadataProviderContext(key, attributes);
+            var context = new ValidationMetadataProviderContext(key, new ModelAttributes(attributes, new object[0]));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DataAnnotationsMetadataProviderTest.cs]: looks new and different.  am I correct calls to the 2-parameter `ModelAttributes` constructor are all for property metadata?

Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DataAnnotationsMetadataProviderTest.cs]: Yes. Where the test is doing something with property metadata, I'm using the constructor that would be called by `GetAttributesForProperty`. It's a bit of a formality and really not needed but I already took the time to make it consistent.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelAttributes.cs (hunk 6 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelAttributes.cs]: #2403

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DataAnnotationsMetadataProviderTest.cs (hunk 8 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DataAnnotationsMetadataProviderTest.cs]: nice

**Comment by dougbu** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/BindingMetadataProviderContext.cs (hunk 7 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/BindingMetadataProviderContext.cs]: This is on a class that's created and then immediately thrown away so I'm not concerned about the extra two fields. Let me know if you think this is important.
